patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,200 @@
++        } else {
++            if (!entity.isHiddenByDefault() || ((CraftEntity) entity).getHandle().tracker.showToEvenWhenHiddenByDefault.contains(this.getHandle().connection)) {
++                return; // Entity isn't hidden
++            }
+         }
++        // Paper end
+         this.hiddenEntities.remove(entity.getUniqueId());
+ 
+         // Paper start
+@@ -1783,9 +1789,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         }
+ 
+         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
+-        if (entry != null && !entry.seenBy.contains(this.getHandle().connection)) {
+-            entry.updatePlayer(this.getHandle());
++        // Paper start - Entity hiddenByDefault
++        if (entry != null) {
++            entry.showToEvenWhenHiddenByDefault.add(this.getHandle().connection);
++
++            if (!entry.seenBy.contains(this.getHandle().connection)) {
++                entry.updatePlayer(this.getHandle());
++            }
+         }
++        // Paper end
+     }
+     // Paper start
+     private void reregisterPlayer(ServerPlayer player) {
+@@ -1843,7 +1855,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public boolean canSee(org.bukkit.entity.Entity entity) {
+-        return !this.hiddenEntities.containsKey(entity.getUniqueId());
++        // Paper start - Entity hiddenByDefault
++        boolean shownWhenHiddenByDefault = true;
++
++        if (entity.isHiddenByDefault()) {
++            ChunkMap.TrackedEntity tracker = ((CraftEntity) entity).getHandle().tracker;
++            shownWhenHiddenByDefault = tracker == null || tracker.showToEvenWhenHiddenByDefault.contains(this.getHandle().connection);
++        }
++
++        return !this.hiddenEntities.containsKey(entity.getUniqueId()) && shownWhenHiddenByDefault;
++        // Paper end
+     }
+ 
+     @Override
+@@ -2892,6 +2913,27 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+     }
+     // Paper end
+ 
++    // Paper start - Entity hiddenByDefault
++    @Override
++    public void setHiddenByDefault(boolean hiddenByDefault) {
++        // Update server lists
++        if (hiddenByDefault) {
++            server.getOnlinePlayers().stream()
++                .filter(entityPlayer -> entityPlayer != this)
++                .filter(entityPlayer -> entityPlayer.canSee(this))
++                .forEach(entityPlayer -> entityPlayer.getHandle().connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.REMOVE_PLAYER, this.getHandle())));
++        } else {
++            server.getOnlinePlayers().stream()
++                .filter(entityPlayer -> entityPlayer != this)
++                .filter(entityPlayer -> !entityPlayer.canSee(this))
++                .forEach(entityPlayer -> entityPlayer.getHandle().connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.ADD_PLAYER, this.getHandle())));
```

Comment:
Streams are cringe and cause unneeded overhead, just use a for loop for this.

Reply:
Done.